src/openrct2/ride/Vehicle.cpp

```diff
@@ -9335,6 +9334,39 @@ void Vehicle::EnableCollisionsForTrain()
     }
 }
 
+void Vehicle::GetSpeedFromTrackElement(TrackElement* trackElement)
+{
+    auto trackType = trackElement->GetTrackType();
+    auto useLegacy = HasUpdateFlag(VEHICLE_UPDATE_FLAG_LEGACY_BOOSTER_SPEED);
+    auto vehicleRTD = GetRide()->GetRideTypeDescriptor();
+    auto trackRTD = get_ride(trackElement->GetRideIndex())->GetRideTypeDescriptor();
+    if (trackType == TrackElemType::PoweredLift
+        || (trackType == TrackElemType::Flat && GetRide()->type == RIDE_TYPE_REVERSE_FREEFALL_COASTER))
+    {
+        brake_speed = DefaultPoweredLiftSpeed; // When PoweredLift gets a speed setting, it will have to choose between default
+                                               // and track's speed based on useLegacy
+        BoosterAcceleration = useLegacy ? vehicleRTD.OperatingSettings.PoweredLiftAcceleration
+                                        : trackRTD.OperatingSettings.PoweredLiftAcceleration;
+        return;
+    }
+    if (TrackTypeHasSpeedSetting(trackType))
+    {
+        auto rawSpeed = trackElement->GetBrakeBoosterSpeed();
+        if (trackType == TrackElemType::Booster && useLegacy)
+        {
+            auto relativeSpeed = trackRTD.GetRelativeSpeed(rawSpeed);
+            brake_speed = vehicleRTD.GetAbsoluteSpeed(relativeSpeed);
+            BoosterAcceleration = vehicleRTD.OperatingSettings.BoosterAcceleration;
+            return;
+        }
+        BoosterAcceleration = trackRTD.OperatingSettings.BoosterAcceleration;
+        brake_speed = rawSpeed;
+        return;
+    }
+    BoosterAcceleration = 0;
+    brake_speed = 0;
```

Comment:
I believe there's potential desyncs if BoosterAcceleration and brake_speed are not set when importing from legacy.